math/cmath/include/algebra/math/impl/cmath_generator.hpp

```diff
@@ -0,0 +1,49 @@
+/** Algebra plugins library, part of the ACTS project
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "algebra/qualifiers.hpp"
+
+// System include(s).
+#include <cassert>
+#include <cstddef>
+#include <type_traits>
+
+namespace algebra::cmath {
+
+/// "Matrix generator", assuming a simple 2D array access
+template <typename size_type, template <typename, size_type> class array_t,
+          typename scalar_t>
+struct matrix_generator {
+
+  /// 2D matrix type
+  template <size_type ROWS, size_type COLS>
+  using matrix_type = array_t<array_t<scalar_t, ROWS>, COLS>;
+
+  // Create zero matrix
+  template <size_type ROWS, size_type COLS>
+  ALGEBRA_HOST_DEVICE inline matrix_type<ROWS, COLS> zero() {
+    return matrix_type<ROWS, COLS>{0};
+  }
+
+  // Create identity matrix
+  template <size_type ROWS, size_type COLS>
+  ALGEBRA_HOST_DEVICE inline matrix_type<ROWS, COLS> identity() {
+    matrix_type<ROWS, COLS> ret{0};
+
+    for (size_type i = 0; i < std::min(ROWS, COLS); ++i) {
```

Comment:
I believe in this setup, unless the compiler is fairly smart (which it may be...), the `std::min(ROWS, COLS)` statement is re-evaluated on every iteration. Just to make the life of the compilers a little easier, you may want to write the loop as:

```c++
const size_type i_max = std::min(ROWS, COLS);
for (size_type i = 0; i < i_max; ++i) {
```